### PR TITLE
New analysis method: diffusion map

### PR DESCRIPTION
**Pull request by euhruska**
Hi everybody,

I implemented a non-linear dimension reduction method, which might be a useful analysis method as it gives slow collective coordinates. I work with diffusion map, so I wanted to make this dimension reduction method easier available for everyone.
An example working with the test data is included. Unfortunately, it's not a great example, since the test data isn't long enough to show any relevant collective slow motions.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] conforms to #719 Analysis interface?

### CODE REVIEW

**Comment by jbarnoud** [comment]
Great to see a new analysis tool!

I did not read the code into details yet, but here are some comments I can already do. Some of them are just nitpicking but it would be good to look at them.

* Could you address the items of the checklist?
* You wrote a documentation for your tool, but it will not appear in the online doc if it is not mentioned in [package/doc/sphinx/source/documentation_pages/analysis_modules.rst](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/doc/sphinx/source/documentation_pages/analysis_modules.rst), and [package/doc/sphinx/source/documentation_pages/analysis](https://github.com/MDAnalysis/mdanalysis/tree/develop/package/doc/sphinx/source/documentation_pages/analysis).
* We are switching to [the numpy way of writing docstring](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt), and it would be good if all new dicstring could follow that format.
* Finally, could you have a look at #719?

I'll try to look at the code latter this week.

**Comment by richardjgowers** [comment]
To answer the question on tests, the numbers generated don't have to be very meaningful scientifically, it just has to catch when the analysis starts producing different results (from the same input!).  So simply running a single frame and recording what the result is today, so that we can check next month we get is the same would be enough (this is a regression test).

**Comment by euhruska** [comment]
I added a test, but it takes 5.795s. I will add another test which is shorter.
Added analysis_modules.rst and analysis/diffusionmap.rst. I'm not sure if I did the numpy way of writing docstring correctly. I wrote the function similar to analysis/align, considering #719.

**Comment by euhruska** [comment]
The faster test takes 2.462s.

**Comment by orbeckst** [comment]
@euhruska also make sure that the email address recorded in your commits matches the one associate with your GitHub account so that commits are properly attributed to you — right now they just show the generic Octocat GitHub logo instead of your GitHub avatar icon – see [Why are my commits linked to the wrong user?](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/)

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+distribution of as protein close to the equilibrium. Further the diffusion map 
+assumes that the diffusion coefficients are constant. The eigenvectors with the 
+largest eigenvalues are the slowest collective coordinates. The complexity of 
+the diffusion map is O(N^2), where N is the number of frames in the trajectory. 
+Instead of a single trajectory a sample of protein structures can be used. 
+The sample should be equiblibrated, at least locally.  Different weights can be used. 
+The order of the sampled structures in the trajectory is irrelevant.
+
+More details how the correst slowest collective coordinates can be calculated.in:
+[1] Nadler, B, Lafon, S, Coifman, R. R, & Kevrekidis, I. G. (2013) Diffusion maps,
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: Make the references a separate section at the end of the background named References and use [reST citations](http://www.sphinx-doc.org/en/stable/rest.html#citations).

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+(in the variables :data:`~MDAnalysis.tests.datafiles.PSF` and
+:data:`~MDAnalysis.tests.datafiles.DCD`). Notice that these test data 
+aren't representing a sample from the equilibrium. This violates a basic 
+assumption of the diffusion map and the results shouldn't be interpreted for this reason. 
+This tutorial shows how to use the diffusionmap function.
+First load all modules and test data ::
+
+   >>> import MDAnalysis
+   >>> import numpy as np
+   >>> import MDAnalysis.lib.qcprot as qcp
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: `qcp` is not needed here

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+   >>> plt.title('diffusion map')
+   >>> plt.xlabel("slowest collective coordinate")
+   >>> plt.ylabel("second slowest collective coordinate")
+   >>> plt.show()
+
+
+Common usage
+------------
+
+To **fit a single structure** with :func:`alignto`::
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: I don't think that this has anything to do with diffusion map:

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+import MDAnalysis.lib.qcprot as qcp
+from MDAnalysis.exceptions import SelectionError, SelectionWarning
+from MDAnalysis.lib.log import ProgressMeter
+import MDAnalysis.analysis.rms as rms
+
+
+logger = logging.getLogger('MDAnalysis.analysis.diffusionmap')
+
+
+def diffusionmap(u, select='all', epsilon='average', k=10, weight=None):
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: Can you put this in the form stipulated by #719 (the "Bauhaus" style): have a `DiffusionMap(base.AnalysisBase)` class? If you can, reduce the `diffusionmap()` function to the bare essentials and call it inside the class.

Hopefuly other people will chime in with hints.

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    """ Non-linear dimension reduction method diffusion map
+    
+    The dimension reduction is done in the following way:
+
+    1. A RMSD between each every pair of frames is calculated.
+    2. The normalized kernel is obtain as in [1] and [2].
+    3. The eigenvalues and eigenvectors of the normalized kernel are the output.
+
+    :Arguments:
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: NumPy style docs, please. See the [Style Guide](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide).

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    if weight is None:
+      weights=np.full((nframes,),1)
+    else:
+      if weight.shape[0]!=nframes:
+        raise ValueError("The weight should have the same length as the trajectroy")
+      else:
+        weights=weight
+
+    for i in range(nframes):
+       logger.info("calculating rmsd from structure "+str(i)+" to all")
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: use `string.format()` formatting

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    the second and higher eigenvectors ev[i+1,:] represent the i-th slowest collective coordinates.
+
+    """
+
+    frames=u.trajectory
+    ref_atoms = u.select_atoms(select)
+    nframes=len(frames)
+    natoms = ref_atoms.n_atoms
+
+    rmsd_matrix=np.zeros((nframes,nframes), dtype=np.float64)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/diffusionmap.py]: Style: follow PEP8, in particular spaces around `=` assignments, space after commas, etc. See links under [Style Guide: Coding Style](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#coding-style).

**Comment by orbeckst** [comment]
Overall looking good but with the recent push towards a more uniform analysis package #719 we'd like you to rewrite it along the new guidelines. Sorry, this is a bit more work, and this was not really clear from the examples that we are currently providing but this is a transitional period where we are trying to rewrite old code and already making new code conform to what is going to be the standard.

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    else:
+      if weight.shape[0]!=nframes:
+        raise ValueError("The weight should have the same length as the trajectroy")
+      else:
+        weights=weight
+
+    for i in range(nframes):
+       logger.info("calculating rmsd from structure "+str(i)+" to all")
+       i_ref=np.copy(u.trajectory[i].positions-ref_atoms.center_of_mass())
+       for j in range(nframes):
```
Comment (jbarnoud) [inline package/MDAnalysis/analysis/diffusionmap.py]: Isn't a RMSD matrix symmetric? Maybe you can calculate only half of the matrix?

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        u = MDAnalysis.Universe(PDB, XTC)
+        eg,ev=diffusionmap.diffusionmap(u, select='backbone', k=5)
+        self.eg=eg
+        self.ev=ev
+       
+
+    def test_eg(self):
+        assert_equal(self.eg.shape, (10,))
+        assert_almost_equal(self.eg[0], 1.0)
+        assert_almost_equal(self.eg[-1],0.03661048812801191)
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: Are that much figures relevant? How precise can we expect the results to be?

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+
+    def test_ev(self):
+        assert_equal(self.ev.shape, (10,10))
+        assert_almost_equal(self.ev[0,0], -0.30796900898350615)
```
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: Same as above.

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+the diffusion map is O(N^2), where N is the number of frames in the trajectory. 
+Instead of a single trajectory a sample of protein structures can be used. 
+The sample should be equiblibrated, at least locally.  Different weights can be used. 
+The order of the sampled structures in the trajectory is irrelevant.
+
+More details how the correst slowest collective coordinates can be calculated.in:
+[1] Nadler, B, Lafon, S, Coifman, R. R, & Kevrekidis, I. G. (2013) Diffusion maps, 
+spectral clustering and reaction coordinates of dynamical systems. Applied and 
+Computational Harmonic Analysis 21, 113–127.
+[2] Rohrdanz, M. A, Zheng, W, Maggioni, M, & Clementi, C. (2013) Determi- nation
```
Comment (jbarnoud) [inline package/MDAnalysis/analysis/diffusionmap.py]: Rogue hyphenation in "Determination".

File: package/MDAnalysis/analysis/diffusionmap.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+   >>> import MDAnalysis.lib.qcprot as qcp
+   >>> import MDAnalysis.analysis.diffusionmap as diffusionmap
+   >>> from MDAnalysis.tests.datafiles import PSF,DCD
+
+In the simplest case, we can simply calculate the diffusion map from one trajectory :func:`diffusionmap`::
+
+   >>> u = MDAnalysis.Universe(PSF,DCD)
+   >>> eg,ev=diffusionmap.diffusionmap(u)
+
+To see how the two slowest collective coordinates how the Other stuff in paper
```
Comment (jbarnoud) [inline package/MDAnalysis/analysis/diffusionmap.py]: I do not understand this sentence.

**Comment by jbarnoud** [comment]
From what I see, you only use the RMSD matrix to calculate epsilon, and you only use one row of the matrix to calculate one value of epsilon. Instead of calculating and storing the whole RMSD matrix, and only then calculating the values of epsilon, could you store only one row of the RMSD matrix at a time, and calculate epsilon in the same loop? This would avoid to store the whole RMSD matrix that can be quite large. This would be you `_single_frame` method in the #719 framework.

Storing only one row of the RMSD matrix at a time makes it difficult to calculate only half of the matrix, though. There is a choice to make: saving CPU time or saving memory.

Expect for this, and the comment above, I see no issue with the code. Though, I am not familiar with the method, and I did not compare the implementation with the papers that are referenced.

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py (hunk 9 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: Truncate to 6 figures or less; float32 machine precision is 1e-7 and you seldom get close. In order for tests to pass reliably you might have decrease precision down to 5. No need to change  the comparison numbers: use the appropriate optional argument (something like precision=5). 

--
Oliver Beckstein
email: orbeckst@gmail.com


Am Mar 7, 2016 um 3:31 schrieb Jonathan Barnoud <notifications@github.com>:

**Comment by euhruska** [comment]
thanks, I will work through the helpful comments.@jbarnoud The main purpose of the RMSD matrix is not epsilon, but getting the eigenvectors and eigenvalues. For that, I need to store the whole matrix. But I will change the code to use the symmetric property to calculate only half of the matrix. Thanks

**Comment by euhruska** [comment]
Why does the Travis CI build fail? Is it related to my code?

**Comment by jbarnoud** [comment]
I don't think so, and it should be fixed with #758.

**Comment by kain88-de** [comment]
@euhruska If you rebase against develop the problem with travis is fixed.

**Comment by jbarnoud** [comment]
Bauhaus is the codename we use to refer to one of the options we discussed in #719. The idea is that an analysis tool inherit from `MDAnalysis.analysis.base.AnalysisBase` and follow the structure imposed by that class. If the analysis or part of it (usually the analysis for a single frame) can be written in a standalone function that does not depend on MDAnalysis, writing that function is encouraged.

The [`Contact` analysis](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/contacts.py#L864) has been rewritten recently to follow the Bauhaus style. [`LinearDensity`](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/lineardensity.py#L32) can also be used as an exemple. These two modules use numpy docstrings.

Note that the interface for analysis tools is not fully defined yet so you may encounter gray areas. What is clear so far is that you need a `_prepare`, a `_conclude`, and a `_single_frame` methods. There seems to be a consensus around a `plot` and a `save` methods.

**Comment by euhruska** [comment]
Is the python3 in Travis build failing because of me?

**Comment by jbarnoud** [comment]
The python 3 build always fails. Do not worry about it. 
I should have a look into that sometime... 
Le 11 mars 2016 3:34 AM, Eugen Hruska <notifications@github.com> a écrit :Is the python3 in Travis build failing because of me?

**Comment by kain88-de** [comment]
@euhruska  our GSoC student @jdetle will implement a PCA during his [project](https://summerofcode.withgoogle.com/projects/#5875821348651008).  Thank you for you work on this feature so far.

**Pull request closed without merging**